rust/lance/src/dataset/write/commit.rs

```diff
@@ -0,0 +1,339 @@
+#[derive(Debug, Clone)]
+pub struct CommitBuilder<'a> {
+    dest: WriteDestination<'a>,
+    use_move_stable_row_ids: Option<bool>,
+    enable_v2_manifest_paths: bool,
+    storage_format: Option<LanceFileVersion>,
+    commit_handler: Option<Arc<dyn CommitHandler>>,
+    store_params: Option<ObjectStoreParams>,
+    object_store_registry: Arc<ObjectStoreRegistry>,
+    object_store: Option<Arc<ObjectStore>>,
+    session: Option<Arc<Session>>,
+    detached: bool,
+    commit_config: CommitConfig,
+}
+
+impl<'a> CommitBuilder<'a> {
+    pub fn new(dest: impl Into<WriteDestination<'a>>) -> Self {
+        Self {
+            dest: dest.into(),
+            use_move_stable_row_ids: None,
+            enable_v2_manifest_paths: false,
+            storage_format: None,
+            commit_handler: None,
+            store_params: None,
+            object_store_registry: Default::default(),
+            object_store: None,
+            session: None,
+            detached: false,
+            commit_config: Default::default(),
+        }
+    }
+
+    /// Whether to use move-stable row ids. This makes the `_rowid` column stable
+    /// after compaction, but not updates.
+    ///
+    /// This is only used for new datasets. Existing datasets will use their
+    /// existing setting.
+    ///
+    /// **Default is false.**
+    pub fn use_move_stable_row_ids(mut self, use_move_stable_row_ids: bool) -> Self {
+        self.use_move_stable_row_ids = Some(use_move_stable_row_ids);
+        self
+    }
+
+    /// Pass the storage format to use for the dataset.
+    ///
+    /// This is only needed when creating a new empty table. If any data files are
+    /// passed, the storage format will be inferred from the data files.
+    ///
+    /// All data files must use the same storage format as the existing dataset.
+    /// If a different format is passed, an error will be returned.
+    pub fn with_storage_format(mut self, storage_format: LanceFileVersion) -> Self {
+        self.storage_format = Some(storage_format);
+        self
+    }
+
+    /// Pass an object store to use.
+    pub fn with_object_store(mut self, object_store: Arc<ObjectStore>) -> Self {
+        self.object_store = Some(object_store);
+        self
+    }
+
+    /// Pass a commit handler to use for the dataset.
+    pub fn with_commit_handler(mut self, commit_handler: Arc<dyn CommitHandler>) -> Self {
+        self.commit_handler = Some(commit_handler);
+        self
+    }
+
+    /// Pass store parameters to use for the dataset.
+    ///
+    /// If an object store is passed, these parameters will be ignored.
+    pub fn with_store_params(mut self, store_params: ObjectStoreParams) -> Self {
+        self.store_params = Some(store_params);
+        self
+    }
+
+    /// Pass an object store registry to use.
+    ///
+    /// If an object store is passed, this registry will be ignored.
+    pub fn with_object_store_registry(
```

Comment:
nit: is there a macro we can use to generate this builder boilerplate?

Reply:
I could maybe write a macro, but doesn't seem that worth it to me.

Reply:
There is https://github.com/colin-kiegel/rust-derive-builder but I agree I'm on the fence about more macros as I think I've developed FUD association between macros and build times.